rust/azure_iot_operations_services/tests/azure_device_registry_client_tests.rs

```diff
@@ -0,0 +1,639 @@
+
+            assert_eq!(updated_asset.name, ASSET_NAME2);
+            assert_eq!(
+                updated_asset.specification.attributes["assetId"],
+                ASSET_NAME2
+            );
+            assert_eq!(updated_asset.status.unwrap(), updated_status);
+
+            // Shutdown adr client and underlying resources
+            assert!(azure_device_registry_client.shutdown().await.is_ok());
+
+            assert!(exit_handle.try_exit().await.is_ok());
+        }
+    });
+
+    assert!(
+        tokio::try_join!(
+            async move { test_task.await.map_err(|e| { e.to_string() }) },
+            async move { session.run().await.map_err(|e| { e.to_string() }) }
+        )
+        .is_ok()
+    );
+}
+
+#[tokio::test]
+// #[ignore]
+async fn observe_asset_update_notifications() {
+    let log_identifier = "observe_asset_update_notifications_network_tests-rust";
+    if !setup_test(log_identifier) {
+        return;
+    }
+    let (session, azure_device_registry_client, exit_handle) =
+        initialize_client(&format!("{log_identifier}-client"));
+
+    let test_task = tokio::task::spawn({
+        async move {
+            // This unobserve is to ensure we start the test with a clean state
+            assert!(
+                azure_device_registry_client
+                    .unobserve_asset_update_notifications(
+                        DEVICE3.to_string(),
+                        ENDPOINT3.to_string(),
+                        ASSET_NAME3.to_string(),
+                        TIMEOUT,
+                    )
+                    .await
+                    .is_ok()
+            );
+
+            let update_desc = format!(
+                "Patch specification update to trigger notification during observe {}",
+                Uuid::new_v4()
+            );
+            let mut observation = azure_device_registry_client
+                .observe_asset_update_notifications(
+                    DEVICE3.to_string(),
+                    ENDPOINT3.to_string(),
+                    ASSET_NAME3.to_string(),
+                    TIMEOUT,
+                )
+                .await
+                .unwrap();
+            log::info!("[{log_identifier}] Asset update observation completed.",);
+
+            let first_notification_notify = Arc::new(Notify::new());
+
+            let receive_notifications_task = tokio::task::spawn({
+                let description_for_task = update_desc.clone();
+                let first_notification_notify = first_notification_notify.clone();
+
+                async move {
+                    log::info!("[{log_identifier}] Asset update notification receiver started.");
+                    let mut count = 0;
+                    loop {
+                        if let Some((asset, _)) = observation.recv_notification().await {
+                            count += 1;
+                            log::info!("[{log_identifier}] Asset Update received: {asset:?}");
```

Comment:
move this inside the `if count ==1`, otherwise you'll print this log twice in the error scenario